File: src/content/docs/topics/cs/coding-problems/greedy/678-valid-parenthesis-string.go
package main

import "fmt"

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func checkValidString(s string) bool {
	// TODO: implement
	return false
}

func runTests() {
	assert(checkValidString("()") == true)
	assert(checkValidString("(*)") == true)
	assert(checkValidString("(*))") == true)
	assert(checkValidString("((") == false)
	assert(checkValidString("*") == true)
	assert(checkValidString("(*") == true)
	assert(checkValidString(")") == false)
	fmt.Println("all tests pass")
}

func main() { runTests() }

File: src/content/docs/topics/cs/coding-problems/greedy/678-valid-parenthesis-string.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function checkValidString(s: string): boolean {
    // TODO: implement
    return false;
}

function _runTests(): void {
    assert(checkValidString('()') === true);
    assert(checkValidString('(*)') === true);
    assert(checkValidString('(*))') === true);
    assert(checkValidString('((') === false);
    assert(checkValidString('*') === true);
    assert(checkValidString('(*') === true);
    assert(checkValidString(')') === false);
    // perf
    const t0 = performance.now();
    checkValidString('(*'.repeat(5000));
    console.log(`perf checkValidString(n=10000): ${(performance.now() - t0).toFixed(1)}ms`);
    console.log('all tests pass');
}

_runTests();
